### PR TITLE
Refactor `DatasetMetrics` and `DiffMetrics`

### LINKED ISSUE #125: Figure out what's lazy and what's not in dask
we can't use the xarray.values property (it will block) - use .data instead
the drop=True subset may also block

in zscore_percent_significant
scipy stats - may not be lazy
also see if np.array() forces to a numpy array

### LINKED ISSUE #190: Feature Request: Decouple computational tasks from plotting tasks
Currently, most of the plotting routines involve doing some computation internally before plotting the results. This approach seems to work fine when dealing with small datasets. However, it becomes a bottleneck when working with large datasets because in some cases we have to recompute the same results to visualize a different slice of the data. Also, the coupling between plotting and computation prevents us from taking full advantage of dask' lazy evaluation philosophy. 

Would it be useful to decouple the plotting tasks from the computational tasks? In other words, can we 
- manage all the computational tasks in standalone objects such as the ones found in `metrics.py`
- Rewrite metrics objects so as to behave like xarray objects which can be extended via accessors: https://xarray.pydata.org/en/stable/internals.html#extending-xarray. With this approach, we would have something along these lines:

```python
metrics_object.ldcpy.plot(......)
```

instead of

```python
ldcpy.plot(col_ds, "TS", sets=["orig"], metric="mean", plot_type="spatial")
```

Right now I don't have all the details for this proposal ironed out, but as I work on #180, I am going to try to put together a minimal example that shows how this would work, and I will post an update here

### PR DESCRIPTION
**Pull request by andersy005**
- Fixes #125 
- Closes #190 
- Closes #180

- [x] Refactor DatasetMetrics by making it an xarray datarray accessor
- [x] Refactor DiffMetrics class
- [ ] Plotting integration
- [ ] Refactor compare_stats utility function

### CODE REVIEW

**Comment by andersy005** [comment]
@pinarda & @allibco, 

I think it's worth getting your feedback on this before moving forward with more changes... I turned the `Metrics` class into an xarray datarray accessor accessible via the `ldc` namespace on a dataarray:

```python
In [1]: import xarray as xr

In [2]: import ldcpy

In [3]: ds = xr.tutorial.open_dataset('air_temperature')

In [5]: metrics = ds.air.ldc(aggregate_dims=['time']).metrics
```

The metrics attribute consists of a Pandas Series will all the metrics... These metrics are computed _once_

```python
In [6]: metrics
Out[6]: 


annual_harmonic_relative_ratio            [[<xarray.DataArray ()>\narray(0.00601499)\nCo...
annual_harmonic_relative_ratio_pct_sig                                                    0
dyn_range                                 <xarray.DataArray 'air' ()>\narray(96.399994, ...
ew_con_var                                [[<xarray.DataArray 'air' ()>\narray(17.121054...
lag1                                      [[<xarray.DataArray 'air' ()>\narray(0.8447127...
lag1_first_difference                     [[<xarray.DataArray 'air' ()>\narray(0.0018499...
mae_day_max                               [[<xarray.DataArray 'dayofyear' ()>\narray(187...
max_abs                                   [[<xarray.DataArray 'air' ()>\narray(279.29, d...
max_val                                   [[<xarray.DataArray 'air' ()>\narray(279.29, d...
mean_abs                                  [[<xarray.DataArray 'air' ()>\narray(260.37564...
mean_squared                              [[<xarray.DataArray 'air' ()>\narray(67795.48,...
mean_val                                  [[<xarray.DataArray 'air' ()>\narray(260.37564...
min_abs                                   [[<xarray.DataArray 'air' ()>\narray(227.59999...
min_val                                   [[<xarray.DataArray 'air' ()>\narray(227.59999...
ns_con_var                                [[<xarray.DataArray 'air' ()>\narray(0.5526597...
odds_positive                             [[<xarray.DataArray 'air' ()>\narray(inf)\nCoo...
pooled_variance                           <xarray.DataArray 'air' ()>\narray(81.94049, d...
pooled_variance_ratio                     [[<xarray.DataArray 'air' ()>\narray(1.684078,...
prob_negative                             [[<xarray.DataArray 'air' ()>\narray(0.)\nCoor...
prob_positive                             [[<xarray.DataArray 'air' ()>\narray(1.)\nCoor...
quantile_val                              [[<xarray.DataArray 'air' ()>\narray(261.29000...
root_mean_squared                         [[<xarray.DataArray 'air' ()>\narray(260.6414,...
standard_deviation                        [[<xarray.DataArray 'air' ()>\narray(11.749104...
standardized_mean                         [[<xarray.DataArray 'air' ()>\narray(-1.279348...
sum_squared                               [[<xarray.DataArray 'air' ()>\narray(5.780513e...
sum_val                                   [[<xarray.DataArray 'air' ()>\narray(760296.9,...
variance                                  [[<xarray.DataArray 'air' ()>\narray(137.99417...
zscore                                    [[<xarray.DataArray 'air' ()>\narray(1197.5319...
zscore_cutoff                                                                         [inf]
zscore_percent_significant                                                              100
dtype: object

```

Subsequent accesses to the metrics returns pre-computed metrics:

```python
In [7]: metrics.zscore
Out[7]: 
<xarray.DataArray 'air' (lat: 25, lon: 53)>
array([[ 1197.5319,  1202.0787,  1200.2444, ...,  1270.1948,  1263.5999,
         1256.8514],
       [ 1217.6824,  1214.032 ,  1204.8011, ...,  1236.7163,  1228.2559,
         1246.5725],
       [ 1150.4156,  1119.2595,  1090.8833, ...,  1197.4948,  1248.0223,
         1348.0845],
       ...,
       [10007.841 , 10314.955 , 10930.48  , ...,  9019.331 ,  8595.423 ,
         8197.951 ],
       [12345.64  , 12675.277 , 12826.182 , ...,  9109.886 ,  8388.199 ,
         7856.7227],
       [15557.243 , 15372.339 , 15522.626 , ...,  9160.014 ,  8689.229 ,
         8342.537 ]], dtype=float32)
Coordinates:
  * lat      (lat) float32 75.0 72.5 70.0 67.5 65.0 ... 25.0 22.5 20.0 17.5 15.0
  * lon      (lon) float32 200.0 202.5 205.0 207.5 ... 322.5 325.0 327.5 330.0
Attributes:
    long_name:     4xDaily Air temperature at sigma level 995
    units:         degK
    precision:     2
    GRIB_id:       11
    GRIB_name:     TMP
    var_desc:      Air temperature
    dataset:       NMC Reanalysis
    level_desc:    Surface
    statistic:     Individual Obs
    parent_stat:   Other
    actual_range:  [185.16 322.1 ]
```

File: ldcpy/dataarray_metrics.py [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+    def __call__(
+        self,
+        aggregate_dims: typing.Union[typing.List, typing.Tuple],
+        grouping: typing.Union[typing.List, typing.Tuple] = None,
+        ddof: int = 1,
+        q: float = 0.5,
+        spre_tol: float = 1.0e-4,
+        time_dim_name: str = 'time',
+        lat_dim_name: str = 'lat',
+        lon_dim_name: str = 'lon',
```
Comment (andersy005) [inline ldcpy/dataarray_metrics.py:73]: Also, I tried removing the hard-coded dimensions names by making them configurable.

**Review by pinarda — COMMENTED** [review]
This looks great so far, thanks Anderson. Is there a stylistic/functional reason to use an accessor instead of a class? I talked about it with Allison, we don't think there is any problem doing it this way and I do like being able to replace the get_metric() methods.

Some of the metrics had optional parameters, for example, q for specifying the quantile we were interested in. Is there a way to do change q and recompute just the quantile without recomputing all the metrics?

**Comment by andersy005** [comment]
Closing this as it's been superseded by other PRs.

**Pull request closed without merging**